SOLR-10703 Add prepare and finish in DocTrasformer

diegoceccarelli opened the pull request:
This patch adds a prepare and a finish method to the interface of DocTransformer allowing a developer to perform actions before/after a doc transformer is applied to a result set. My use case was to benchmark the performance of a transformer, since transformer time is not part of QTime.

m-khl commented on solr/contrib/ltr/src/java/org/apache/solr/ltr/response/transform/LTRFeatureLoggerTransformerFactory.java:
```diff
@@ -187,8 +187,8 @@ public String getName() {
     }
 
     @Override
-    public void setContext(ResultContext context) {
```
here is the problem: we have a plenty of transformers they are not changed?  (assuming they shouldn't be. see below)

diegoceccarelli replied:
Hey @m-khl  thanks for reviewing my code! 

> here is the problem: we have a plenty of transformers they are not changed? (assuming they shouldn't be. see below)

In the main abstract class I only deprecated `DocTransformer::setContext`, so they will still compile. I see your point, and I could change so that `prepare` calls also `setContext`, in this way the behaviour of existing transformers will be preserved

m-khl commented on solr/core/src/java/org/apache/solr/handler/component/RealTimeGetComponent.java:
```diff
@@ -1210,7 +1211,7 @@ public SolrQueryRequest getRequest() {
     }
     
     /** @returns null */
-    public Iterator<SolrDocument> getProcessedDocuments() {
```
not clear why

m-khl commented on solr/core/src/java/org/apache/solr/handler/component/RealTimeGetComponent.java:
```diff
@@ -296,13 +297,13 @@ public void process(ResponseBuilder rb) throws IOException
        docList.add(doc);
      }
-
+     if ( null != transformer) transformer.finish();
```
if it impacts RTG it also should be tested, otherwise, it can lead to leak from some heavy transformer

diegoceccarelli replied:
I'll take a look, thank you

m-khl commented on solr/core/src/java/org/apache/solr/response/BinaryResponseWriter.java:
```diff
@@ -119,11 +120,15 @@ public boolean wantsAllFields() {
 
     protected void writeResultsBody( ResultContext res, JavaBinCodec codec ) throws IOException {
       codec.writeTag(JavaBinCodec.ARR, res.getDocList().size());
-      Iterator<SolrDocument> docStreamer = res.getProcessedDocuments();
```
I can't spot the difference between these lines, are there?

m-khl commented on solr/core/src/java/org/apache/solr/response/DocsStreamer.java:
```diff
@@ -142,6 +142,11 @@ public boolean hasNext() {
     return docIterator.hasNext();
   }
 
+  // called at the end of the stream
```
can't we leverage Closeable here and get some sugar&warns? 
Also, line 89 still calls setContext() .. is it right? or I'm missing something?

diegoceccarelli replied:
the reason for `prepare` and `finish` was to have the same naming of components, imo it makes easier to understand what's going on, I do like the Closeable idea, do you think it would be better?

> Also, line 89 still calls setContext() .. is it right? or I'm missing something?

No, I missed line 89, thank you

m-khl commented on solr/core/src/java/org/apache/solr/response/transform/DocTransformer.java:
```diff
@@ -41,12 +41,26 @@
+   * @param context The {@link ResultContext} stores information about how the documents were produced.
+   * @see #needsSolrIndexSearcher
+   */
+  public void prepare( ResultContext context ) {
```
why the existing setContext() is not enough?

diegoceccarelli replied:
if you want to do something different than setting the context (see LTRFeatureLoggerTransformerFactory for example), the name is misleading. we could keep the setContext and just add the `prepare` method.

m-khl replied:
I suppose if anything can be done with existing facilities, they should be utilised. I don't take it much if some name is misleading. We have a lot of much more severe issues and live with them.  -1 for "just add"

m-khl commented on solr/core/src/java/org/apache/solr/response/transform/DocTransformers.java:
```diff
@@ -88,4 +89,11 @@ public boolean needsSolrIndexSearcher() {
+  @Override
+  public void finish() {
+    for( DocTransformer a : children ) {
+      a.finish();
```
what is one of them throws the exception shouldn't it continue to finish others?

diegoceccarelli replied:
yes? I think that if a transformer throws an exception solr request should fail (you can still try and catch the exception inside the transformer if you want to obtain 'silent' behaviour) what do you think?

m-khl replied:
I don't think it a way to design SPI (not A). Such silence should be either enforced with final/abstract constraints; or enforced in runtime. Imagine if some good transformer needs finish() to release a resource, but there is some badass sibling one. and it's enbaled from time-to-time at prod. requests. The instance will steadily leak, and it would be hard to find out why.

m-khl commented on solr/core/src/java/org/apache/solr/response/transform/DocTransformer.java:
```diff
@@ -93,6 +107,12 @@ public void setContext( ResultContext context ) {
+  /**
+   * This is called after a transformer has been applied to all the documents in the results set
+   */
+  public void finish(){
```
is it possible to assert that finish is called always during the tests on the base class?

diegoceccarelli replied:
good idea, I'll take a look

m-khl commented:
Here is the overall note: Transformers has already setContext, thus they can already initialize. Finish logic can be implemented with SolrRequestInfo.addCloseHook(Closeable)

diegoceccarelli commented:
@m-khl  I updated the patch, highlights: 
   1. I removed `prepare` and relied on the old `setContext()`;
   2. now a DocTransformer implements Closeable and provides the method `close()`
   3. If a transformer fails it doesn't affect the other trasformers (+ Test)

TODO:
I didn't find a way to check that DocTrasformer::close is always called in the tests, any idea? 

The `close()` is called from `DocsStreamer` so RealTimeGetRequest component should not be affected by this change, do you think it should? 






minor: I think we should rename the Jira item in something like `Make DocTransformer Closeable` or something like that

dsmiley commented:
Maybe this should use Solr's `IOUtils.closeQuietly`?  It's tempting to use Lucene's `IOUtils.closeWhileHandlingException` since it'll save us from even looping here, although that one doesn't log the exception; it's swallowed completely :-/

michaelbraun commented:
DocTransformer post-implementing Closeable has a close() method but it is not annotated @Override - it should probably have

diegoceccarelli commented:
Thanks @dsmiley  and @michaelbraun, I updated the code

tflobbe commented on solr/core/src/test/org/apache/solr/response/TestCustomDocTransformer.java:
```diff
@@ -68,6 +69,51 @@ public void testCustomTransformer() throws Exception {
+
+    // finish() will double the number of documents
+    assertEquals(max*2, CustomFinishTransformerFactory.finishTrasformer.counter);
+    CustomFinishTransformerFactory.finishTrasformer.counter = 0;
```
Is this needed? Shouldn't setContext() set this value?

tflobbe commented on solr/core/src/test/org/apache/solr/response/TestCustomDocTransformer.java:
```diff
@@ -125,4 +171,86 @@ public static String getAsString(String field, SolrDocument doc) {
+
+    @Override
+    public void transform(SolrDocument doc, int docid) throws IOException {
+      counter++;
```
Suggestion. Throw exception if close() was already called in this request

tflobbe commented on solr/core/src/test/org/apache/solr/response/TestCustomDocTransformer.java:
```diff
@@ -68,6 +69,51 @@ public void testCustomTransformer() throws Exception {
+        "rows", String.valueOf(max),
+        "wt", "javabin"
+    ));
+    assertEquals(max*2, CustomFinishTransformerFactory.finishTrasformer.counter);
```
Suggestion: Assert that the exception was actually thrown, and add a "ignoreException" to not pollute the logs

diegoceccarelli commented on solr/solrj/src/java/org/apache/solr/common/util/IOUtils.java:
```diff
@@ -31,7 +32,7 @@ public static void closeQuietly(Closeable closeable) {
       }
     } catch (Exception e) {
-      LOG.error("Error while closing", e);
+      SolrException.log(LOG, "Error while closing "+closeable.getClass());
```
I had to modify the logging here in order to have `ignoreException` working in the tests.

diegoceccarelli commented:
@tflobbe @m-khl ping :)
